jax/interpreters/xla.py

```diff
@@ -463,22 +464,88 @@ def jaxpr_collectives(jaxpr):
 
 ### xla_call underlying jit
 
-def _xla_call_impl(fun: lu.WrappedFun, *args, device, backend, name):
-  compiled_fun = _xla_callable(fun, device, backend, name, *map(arg_spec, args))
+def _xla_call_impl(
+    fun: lu.WrappedFun,
+    *args,
+    device: Optional[xe.Device],
+    backend,
+    name: str,
+    donate: Sequence[bool],
+):
+  if len(args) != len(donate):
+    if any(donate):
+      raise ValueError("Taking the JVP of a jit/pmap decorated function which "
+                       "donates some arguments is not supported. Maybe try "
+                       "`jit(partial(jvp, f), donate_argnums=...)((x,), (x,))` "
+                       "instead of "
+                       "`jvp(jit(f, donate_argnums=...), (x,), (x,))`")
+    donate = (False,) * len(args)
+
+  arg_specs = safe_map(arg_spec, args, donate)
+  compiled_fun = _xla_callable(fun, device, backend, name, *arg_specs)
   try:
     return compiled_fun(*args)
   except FloatingPointError:
     print("Invalid value encountered in the output of a jit function. "
           "Calling the de-optimized version.")
     return fun.call_wrapped(*args)  # probably won't return
 
+def flatten_shape(s: xe.Shape) -> Sequence[Tuple[Sequence[int], xe.Shape]]:
```

Comment:
Claim: JAX only uses flat shapes these days. We use tuples for exactly two reasons:
a) to pass more than N arguments to a computation, where N is a few hundred, and
b) as return values from computations (we may even destructure these before they make it back to JAX.)

Do we still need this function?

Reply:
I think its good to have this function to handle that case, my preference would be to keep the general implementation since it is not all that complex. I guess we could shave a few lines off if we made this switch on tuple_args. Please feel free to reopen if you feel strongly.